### PR TITLE
Add streaming session and ability to use streaming

### PR DESCRIPTION
**Pull request by Krivoblotsky**
## What

1. Chats and Completions have been updated with streaming functionality
2. ChatStreamingResult has been created
3. Tests have been updated

## Why

#14 

## Affected Areas

Completions, Chats

### CODE REVIEW

File: Sources/OpenAI/OpenAI.swift [hunk 1]
```diff
@@ -35,6 +35,7 @@ final public class OpenAI: OpenAIProtocol {
     }
     
     private let session: URLSessionProtocol
+    private var streamingSessions: [NSObject] = []
```
Comment (DJBen) [inline Sources/OpenAI/OpenAI.swift:38]: You may want [`NSHashTable`](https://developer.apple.com/documentation/foundation/nshashtable) to give it set semantics and broader memory operations.

File: Sources/OpenAI/OpenAI.swift [hunk 2]
```diff
@@ -64,7 +65,11 @@ final public class OpenAI: OpenAIProtocol {
     }
     
     public func chats(query: ChatQuery, completion: @escaping (Result<ChatResult, Error>) -> Void) {
-        performRequest(request: JSONRequest<ChatResult>(body: query, url: buildURL(path: .chats)), completion: completion)
+        if query.stream == true {
```
Comment (DJBen) [inline Sources/OpenAI/OpenAI.swift:68]: nit: query.stream

File: Sources/OpenAI/Public/Models/ChatResult.swift [hunk 3]
```diff
@@ -10,13 +10,20 @@ import Foundation
...
     public struct Choice: Codable, Equatable {
+        public struct Delta: Codable, Equatable {
+            public let content: String?
+            public let role: Chat.Role?
+        }
+        
         public let index: Int
-        public let message: Chat
-        public let finishReason: String
+        public let message: Chat?
```
Comment (DJBen) [inline Sources/OpenAI/Public/Models/ChatResult.swift:19]: If you want to mix the complete and partial results, can you comment in what condition these properties are `nil`?

I have moderate preference in separating a complete `Choice` from a `PartialChoice`, provided that they have major differences, despite the official OpenAI API mix them together.

```swift
// Choice stays the same

public struct PartialChoice: Codable, Equatable {
  public let index: Int
  public let delta: Delta
}

```

**Comment by DJBen** [comment]
Since we are using two separate functions for the `chats(query:)` and `chatStream(query:)` to represent streaming and non-streaming version of the chat api, we can potentially get rid of the `stream: Bool?` property inside `ChatQuery`, and modify the encoder to serialize `stream: false` and `stream: true` respectively when calling these two different functions. But it's probably a large model change.

**Comment by DJBen** [comment]
I added streaming in the demo app #59

**Review by DJBen — APPROVED** [review]

**Comment by Krivoblotsky** [comment]
It looks like everything is working as expected 🤞 
Going to check more tests on weekends and proceed. 
https://user-images.githubusercontent.com/1411778/235212448-0aa844d5-7ffb-487e-a0bb-37eed5666f48.mov

File: Sources/OpenAI/OpenAI.swift [hunk 4]
```diff
@@ -127,7 +132,26 @@ extension OpenAI {
...
+            let request = try request.build(token: configuration.token, organizationIdentifier: configuration.organizationIdentifier, timeoutInterval: configuration.timeoutInterval)
+            let session = StreamingSession<ResultType>(urlRequest: request)
+            session.onReceiveContent = {_, object in
+                completion(.success(object))
+            }
+            session.onProcessingError = {_, error in
+                completion(.failure(error))
+            }
+            session.onComplete = { [weak self] object, error in
+                self?.streamingSessions.removeAll(where: { $0 == object })
```
Comment (tikimcfee) [inline Sources/OpenAI/OpenAI.swift:149]: Heya! Jumping in, I got a little excited to test this out. Doing some testing myself, this needs to call `completion` in some way in order for the backing `AsyncStream` to actually finish. At the moment I'm throwing an error like this:
```
enum StreamError: Error {
 case streamStop
}
//...
completion(.failure(StreamError.streamStop)
```

Which allows the `throw` to act as the trigger to stop. There's likely a more elegant way to do this, so hopefully the report at least helps track that part down. Cheers, and thanks much for your time!

**Comment by DJBen** [comment]
Hi @Krivoblotsky do you know when this can be merged?

**Comment by Krivoblotsky** [comment]
Hi, @DJBen 
I have plans for this week for this to be updated and merged.

File: Sources/OpenAI/Public/Models/ChatResult.swift (hunk 3 above)
Comment (Krivoblotsky) [reply Sources/OpenAI/Public/Models/ChatResult.swift:20]: I've decided to keep `ChatResult`to not break backward compatibility and created "ChatStreamingResult" along with a separate function for streaming. I found this solution more straightforward and clean. What do you think about this?

**Comment by Krivoblotsky** [comment]
Hey, @DJBen. 
I'm ready to merge it, but wanted to ask you for a review first 🙏

**Review by DJBen — APPROVED** [review]
LG!

**Pull request merged**